### PR TITLE
tests: split DDP & common + running xdist

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Identify DDP tests and split testing into two steps:
- run DDP in a single thread as it was until now
- the rest of the tests try to run in parallel

In addition move unitest steps back to the main workflow

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2127.org.readthedocs.build/en/2127/

### CODE REVIEW

**Comment by Borda** [comment]
Seems like this does not even start properly, could be because of:
https://github.com/Lightning-AI/torchmetrics/blob/886c09b01423a5013f289e95672373bfbc33de85/tests/unittests/conftest.py#L55-L73
cc: @SkafteNicki

**Comment by SkafteNicki** [comment]
> Seems like this does not even start properly, could be because of?

Yeah that would definitely be the problem, if I remember correctly you cannot start a child process inside a child process. So I am not sure if we will be able to use xdist for testing.

**Comment by SkafteNicki** [comment]
I can get it to run locally, but it is complaining with error
```
AttributeError: module pytest has no attribute pool
```
meaning that we need to refactor the code where we initialize the pool used for multiprocessing testing if we want this PR do land.

File: .github/workflows/ci-integrate.yml [hunk 1]
```diff
@@ -56,6 +56,10 @@ jobs:
         if: matrix.requires == 'oldest'
         run: python .github/assistant.py set-oldest-versions --req_files='["requirements/_integrate.txt"]'
 
+      - name: Freeze PIL (hotfix)
+        # import of PILLOW_VERSION which they recently removed in v9.0 in favor of __version__
+        run: pip install "Pillow<9.0" # It messes with torchvision
+
```
Comment (Borda) [inline .github/workflows/ci-integrate.yml:62]:
```suggestion
```

File: .github/workflows/ci-tests.yml [hunk 2]
```diff
@@ -103,6 +104,11 @@ jobs:
         # NOTE: run coverage on tests does not propagate failure status for Win, https://github.com/nedbat/coveragepy/issues/1003
         run: python -m pytest torchmetrics --reruns 3 --reruns-delay 2
 
+      - name: Freeze PIL (hotfix)
+        if: ${{ matrix.python-version != '3.11' }}
+        # import of PILLOW_VERSION which they recently removed in v9.0 in favor of __version__
+        run: pip install "Pillow<9.0" # It messes with torchvision
+
```
Comment (Borda) [inline .github/workflows/ci-tests.yml:111]:
```suggestion
```

**Comment by Borda** [comment]
this is stated from master:
image
so seems we get **15-40min boost** on GPU
image

**Review by SkafteNicki — APPROVED** [review]
Fantastic work @Borda

File: pyproject.toml [hunk 3]
```diff
@@ -19,6 +19,9 @@ addopts = [
     "--color=yes",
     "--disable-pytest-warnings",
 ]
+markers = [
+    "DDP: mark a test as DDP",
```
Comment (Borda) [inline pyproject.toml:23]:
```suggestion
    "DDP: mark a test as Distributed Data Parallel",
```

**Comment by lantiga** [comment]
Great job @Borda ⚡️

**Comment by stancld** [comment]
@Borda Wuaah, this is amazing! 🥇

**Pull request merged**